lua/telescope/health.lua

```diff
@@ -0,0 +1,115 @@
+
+local optional_dependencies = {
+  {
+    finder_name = "live-grep",
+    package = {
+      {
+        name = "rg",
+        url = "[BurntSushi/ripgrep](https://github.com/BurntSushi/ripgrep)",
+        optional = false,
+      },
+    },
+  },
+  {
+    finder_name = "find-files",
+    package = {
+      {
+        name = "fd",
+        url = "[sharkdp/fd](https://github.com/sharkdp/fd)",
+        optional = true,
+      },
+    },
+  },
+}
+
+local required_plugins = {
+  { lib = "popup", optional = false },
+  { lib = "plenary", optional = false },
+  {
+    lib = "nvim-treesitter",
+    optional = true,
+    info = "Install [nvim-treesitter](https://github.com/nvim-treesitter/nvim-treesitter) to enable preview highlight capabilities.",
+  },
+}
+
+local check_binary_installed = function(package)
+  local file_extension = jit.os == "Windows" and "exe" or ""
+  local filename = package.name .. file_extension
+  if fn.executable(filename) == 0 then
+    return
+  else
+    local handle = io.popen(filename .. " --version")
+    local binary_version = handle:read "*a"
+    handle:close()
+    return true, binary_version
+  end
+end
+
+local function lualib_installed(lib_name)
+  local res, _ = pcall(require, lib_name)
+  return res
+end
+
+local M = {}
+
+M.check_health = function()
+  -- Required lua libs
+  health_start "Checking for required plugins"
+  for _, plugin in ipairs(required_plugins) do
+    if lualib_installed(plugin.lib) then
+      health_ok(plugin.lib .. " installed.")
+    else
+      local lib_not_installed = plugin.lib .. " not found."
+      if plugin.optional then
+        health_warn(("%s %s"):format(lib_not_installed, plugin.info))
+      else
+        health_error(lib_not_installed)
+      end
+    end
+  end
+
+  -- external dependencies
+  -- TODO: only perform checks if user has enabled dependency in their config
+  health_start "Checking external dependencies"
+
+  for _, opt_dep in pairs(optional_dependencies) do
+    for _, package in ipairs(opt_dep.package) do
+      local installed, version = check_binary_installed(package)
+      if not installed then
+        local err_msg = ("%s: not found."):format(package.name)
+        if package.optional then
+          health_warn(("%s %s"):format(err_msg, ("Install %s for extended capabilities"):format(package.url)))
+        else
+          health_error(
+            ("%s %s"):format(
+              err_msg,
+              ("`%s` finder will not function without %s installed."):format(opt_dep.finder_name, package.url)
+            )
+          )
+        end
+      else
+        local eol = version:find "\n"
+        health_ok(("%s: found %s"):format(package.name, version:sub(0, eol - 1) or "(unknown version)"))
+      end
+    end
+  end
+
+  -- Extensions
+  health_start "Installed extensions"
+  local installed = {}
+  for extension_name, _ in pairs(extension_info) do
+    installed[#installed + 1] = extension_name
+  end
```

Comment:
hmm what if extensions could hook into checkhealth? by optionally implementing a function? For example we then could check if everything is working (fzf calc something)

Reply:
I've implmented that. If an extension has exposed a `health` function, the parent Telescope healthcheck will call it.
All responsiblity of calling health_ok/health_error, etc is down to the extension, ie. no sanitization of what goes on in that function occurs.

Reply:
Okay cool thanks :) i'll try it tomorrow with fzf-native and give a overall better feedback but right now code looks good :)